### ISSUE TITLE
Possible compiler bug with pointer variable vs direct pointer

### ISSUE THREAD

**Issue by clbr** [cc65/cc65#218]
This file:
```
#include <nes.h>

int main() {

        unsigned char *ppuaddr = (unsigned char *) 0x2006;
        unsigned char *ppudata = (unsigned char *) 0x2007;
        unsigned char *ppuctrl = (unsigned char *) 0x2000;
        unsigned char *ppumask = (unsigned char *) 0x2001;
        unsigned char *ppuscroll = (unsigned char *) 0x2005;

        waitvblank();

        *((unsigned char*)0x2006) = 0x3F;
        *((unsigned char*)0x2006) = 0x00;
//      *((unsigned char*)0x2007) = 1;
        *ppudata = 1; // uncomment the above and comment this -> it works

        *ppuscroll = 0;
        *ppuscroll = 0;

        *ppumask = 8;

        while (1);
        return 0;
}
```

fails when run - the background color is grey. If I comment the *ppudata assignment and uncomment the direct pointer assignment, it works - the bg color is blue.

I don't speak enough asm to say why the pointer-var fails and direct pointer works. Here is the diff of the assemblies:
```
--- wrong.s     2015-10-05 22:33:49.000000000 +0300
+++ right.s     2015-10-05 22:33:43.000000000 +0300
@@ -46,13 +46,9 @@
        ldx     #$00
        lda     #$00
        sta     $2006
-       ldy     #$07
-       jsr     ldaxysp
-       jsr     pushax
        ldx     #$00
        lda     #$01
-       ldy     #$00
-       jsr     staspidx
+       sta     $2007
        ldy     #$01
        jsr     ldaxysp
        jsr     pushax

```

**Comment by clbr** [comment]
Enabling optimizations, either at -O or -Oisr level, does not change anything. Right still works, wrong still fails even on full optimization, even if pointers are declared as "unsigned char * const", at which level the compiler should be able to optimize them to direct calls.

Still, a missing optimization doesn't explain the failure.

**Comment by oliverschmidt** [comment]
Rather independent from the question of a potential compiler bug you're much better of with the direct assignment. `ppudata` and friends are pointer variables while ` *((unsigned char*)0x2007)`and friends are pointer constants. Using variables is in general slower/larger than using constants. Think of
```
int x = a;
```
versus
```
int x = 1;
```
For pointers there's much more about it with cc65: Using a pointer variable generally means to load its value, store it into two zero page addresses and the use 6502 "indirect adressing" to access the pointer. Using a pointer constant is in contrast explicitly optimized in cc65 to "just" access the address given in the constant.

**Comment by clbr** [comment]
True, and that's how I will work around it.

However, a
unsigned char * const ptr = (unsigned char *) 0x400;

is a constant pointer, and should be equivalent to directly using the
value - a missing optimization, as mentioned above. Using variables
instead of macros comes with the usual advantages in type safety,
and a lack of evaluation order gotchas.

**Comment by oliverschmidt** [comment]
@clbr: The code you posted above doesn't contain the `const` you mention now and my comment was about your code above ;-) Apart from that is cc65 an open source project and you're invited to improve the compiler in a way that the "should" in your last comment becomes true ;-))

**Comment by spiro-trikaliotis** [comment]
Hello,

* On Tue, Oct 06, 2015 at 03:30:49AM -0700 clbr wrote:
 
> However, a
> unsigned char * const ptr = (unsigned char *) 0x400;
> 
> is a constant pointer, and should be equivalent to directly using the
> value

Well - it depends.

Note that a "constant" in C is not what many people believe it to be,
cf. for example http://c-faq.com/ansi/constasconst.html.

Even with your example, there are things you can do with it that might
not be obvious:

        unsigned char * const ptr = (unsigned char *) 0x400;

        unsigned char * const * p1 = &ptr;
        unsigned char *       * p2 = (unsigned char **) p1;

        *p1 = 0x600;

Ouch, you have modified your const pointer!

Of course, this example is artificial, but it shows one thing: In order
to treat a const pointer (or any const element in C) as really constant,
a dataflow analysis has to be done. Otherwise, the compiler cannot
assume constness.

Take another example: Your ptr is a parameter to your function; outside
the function, the const does not exist, and you have some concurrent
things (multitasking, an interrupt routine) that modifies the pointer.

Bottom line: There is almost (?) never a reason to declare a pointer as
constant. If the compiler does a dataflow analysis, it can treat it as
const out of its own, without the help of the programmer. If it cannot
deduce this, the "const" qualifier does not help it, either.


tl;dr: The qualifier "const" means read-only in C, not a constant (as in
C++).

Regards,
Spiro.

-- 
Spiro R. Trikaliotis
http://www.trikaliotis.net/

**Comment by clbr** [comment]
> Well - it depends.

Yes, I know. Flow analysis is expected of any compiler, but I infer
that cc65 does not have any (yet?).

An easy solution without complex analysis would be a mode that forbids
removing const with casts.

> Take another example: Your ptr is a parameter to your function; outside
> the function, the const does not exist, and you have some concurrent
> things (multitasking, an interrupt routine) that modifies the pointer.

The compiler is allowed to ignore any multitasking, and work with what
the programmer tells it. If I tell it the pointer will not change, it
will not change, and it can rely on the contract.

> Bottom line: There is almost (?) never a reason to declare a pointer as
> constant. If the compiler does a dataflow analysis, it can treat it as
> const out of its own, without the help of the programmer. If it cannot
> deduce this, the "const" qualifier does not help it, either.

The difference is that by declaring it const, you're protecting yourself
from making mistakes with it. In the absence of LTO, it can also help
optimization, as whole-program flow analysis can't be done then.

It can also affect whether static variables can be placed in rodata or
not.

**Comment by greg-king5** [comment]
There are two questions in this issue:

1. Why do two ways to write into an I/O register compile to different machine language?
2. Why does one of those ways fail?

Spiro answered the first question.  I will answer the second one.

It is a conflict between a bug in the 6502 chip and the way that PPU I/O works.

The C expression "`*ppudata = 1;`" compiles to code that ends with "`sta (ptr1),y`".
During the execution of that instruction, the Y index register is added to the target address.  It happens during an entire machine cycle.  But, the chip's address and data busses are not idle; they always are doing something -- writing something or reading something.  In this case, the chip writes (something) to `$2007` while it is adding that index; and then, it writes (again) the `$01` to `$2007`.

When the CPU writes a value into `$2007` (the Video RAM data port), the PPU automatically increments the VRAM address (from register `$2006`).  Therefore, the 6502 bug actually puts the `$01` into the location that is one byte beyond the desired place.

You can see that problem in action by using an address that is one lower than where you want it to be:

```C
#include <nes.h>

void main(void) {
    unsigned char *ppuaddr = (unsigned char *) 0x2006;
    unsigned char *ppudata = (unsigned char *) 0x2007;
    unsigned char *ppuctrl = (unsigned char *) 0x2000;
    unsigned char *ppumask = (unsigned char *) 0x2001;
    unsigned char *ppuscroll = (unsigned char *) 0x2005;

    waitvblank();

    *((unsigned char*)0x2006) = 0x3E;  // 0x3F00 - 1
    *((unsigned char*)0x2006) = 0xFF;
    *ppudata = 1;

    *ppuscroll = 0;
    *ppuscroll = 0;
    *ppumask = 8;

    while (1) ;
}
```

You will see blue.

**Comment by oliverschmidt** [comment]
> [...] the PPU automatically increments the VRAM address [...]

Unfortunately I don't understand the C64 well enough to know that there's an auto increment register :-(

It's "of course" a rather bad idea to deal with auto increment registers from C !

Im doing it neverless here for demonstration purposes: https://github.com/oliverschmidt/ip65/blob/master/supplement/w5100.c

There you find
```
// Both pragmas are obligatory to have cc65 generate code
// suitable to access the W5100 auto-increment registers.
#pragma optimize      (on)
#pragma static-locals (on)
```
and instances of
```
    // The variables are necessary to have cc65 generate code
    // suitable to access the W5100 auto-increment registers.
    [...]
```

So you can "trick" cc65 into generating code suitable for auto-increment registers with the "right" source code combined with the "right" compiler flags but you need to be very careful and check the generated code after every source code change...

**Comment by oliverschmidt** [comment]
As far as I can see everything is clarified by now. Thanks to the contributors :-)

**Comment by clbr** [comment]
On Wed, 07 Oct 2015 05:54:20 -0700
Oliver Schmidt <notifications@github.com> wrote:

> Closed #218.

Hey, this should at least be documented in the nes page before closing.

**Comment by greg-king5** [comment]
Another thing that needs to be done is that the PPU and APU chips need to have structured register definitions added to "`<nes.h>`".

(If they had been there, then Clbr would not have noticed these issues.)  ;-)

**Comment by oliverschmidt** [comment]
> Hey, this should at least be documented in the nes page before closing.

I won't reject any pull request improving the documentation in this - or in fact any other - area.

Apart from that https://github.com/cc65/wiki/wiki/PEEK-and-POKE already says

"Please note, that the addresses used must not necessarily be constant (as in the examples above). However, while the code generated by the compiler for constant addresses is quite good (they get translated into a series of load and store instructions), for non constant addresses a subroutine is called, so this is a lot slower."

Given that the auto-incrementing register discussed here is indeed placed at a fixed address I don't see why anybody wouldn't want the more efficient code. And that code doesn't cause issues.
